rules/execute-command.js

```diff
@@ -132,33 +131,39 @@ async function runCommand(command, args = []) {
   }
 }
 
-function smuggleLockfile(rootDir) {
+function smuggleYarnProjectFiles(rootDir) {
   // Support non-hermetic yarn.lock/.pnp.cjs/.pnp.loader.mjs smuggling into sandbox
   // This allows a repo to use yarn plugins to implement custom change detection
   // to avoid invalidating top-level cache in cases where yarn.lock is touched but only affects certain projects
   // See: Lockfile delegation section in README
 
-  try {
-    const realRoot = dirname(realpath(`${rootDir}/package.json`)); // FIXME: technically a target may not depend on this file, so it's not actually guaranteed to exist, even though it usually does
+  // @see {@link https://github.com/yarnpkg/berry/blob/07cf3531002a3f25fd6309e05d0cf1b233c59cd4/packages/yarnpkg-fslib/sources/path.ts#L21-L35}
+  const yarnProjectFiles = [
+    'yarn.lock',
+    '.pnp.data.json',
+    '.pnp.loader.mjs',
+    '.pnp.cjs',
+    '.yarnrc.yml',
+    '.yarnrc',
+    'package.json',
+  ];
 
-    //only smuggle if needed
-    const yarnLock = `${rootDir}/yarn.lock`;
-    const realYarnLock = `${realRoot}/yarn.lock`;
-    if (!exists(yarnLock) && exists(realYarnLock)) {
-      symlink(realYarnLock, yarnLock, 'file');
-    }
-    const pnpCjs = `${rootDir}/.pnp.cjs`;
-    const realPnpCjs = `${realRoot}/.pnp.cjs`;
-    if (!exists(pnpCjs) && exists(realPnpCjs)) {
-      symlink(realPnpCjs, pnpCjs, 'file');
-    }
-    const pnpLoader = `${rootDir}/.pnp.loader.mjs`;
-    const realPnpLoader = `${realRoot}/.pnp.loader.mjs`;
-    if (!exists(pnpLoader) && exists(realPnpLoader)) {
-      symlink(realPnpLoader, pnpLoader, 'file');
+  try {
+    const realRootDir = dirname(realpath(`${rootDir}/WORKSPACE`));
+    for (const file of yarnProjectFiles) {
+      try {
+        // only smuggle if needed
+        const filepath = `${rootDir}/${file}`;
+        const realFilepath = `${realRootDir}/${file}`;
+        if (!exists(filepath) && exists(realFilepath)) {
+          symlink(realFilepath, filepath, 'file');
+        }
+      } catch (e) {
+        // smuggling failed, assume a file already exists and keep going
+      }
     }
   } catch (e) {
-    // smuggling failed, assume yarn.lock exists and keep going
+    // smuggling failed, assume required files are already imported into bazel sandbox
```

Comment:
I left it just in case the fs call on L152 throws an exception (e.g., in case the `WORKSPACE` is not present in the sandbox?): https://github.com/uber-web/jazelle/blob/2177fb908199528c7c16652bb77fe8aac5d29611/rules/execute-command.js#L151-L152

Unless you think it'd be better to treat such error as fatal exception?

Reply:
yeah if WORKSPACE is not there, stuff is going to break way earlier than this

Reply:
Hm.. We may need to find some better way to infer the project root directory from a sandbox environment. Cuz it seems like having a `WORKSPACE` in a depset does more harm than is necessary. E.g., a simple change like adding a whitespace in this file invalidates all targets that explicitly included it in the depset, as opposed to the standard bazel's build analysis which would invalidate changed targets in a more granular fashion, detecting any changes to the underlying rules / actions..